### PR TITLE
Hashring fixes

### PR DESCRIPTION
**Pull request by thomasst**
* Use pubsub.get_message() instead of socket hack to get new messages
* Compatibility with different redis-py decode_responses settings

### CODE REVIEW

**Comment by tsx** [comment]
Is this still an issue?

**Comment by thomasst** [comment]
Yeah, it's low priority but we should resolve conflicts & still merge this.

**Comment by jpmelos** [comment]
`dialer` has been running with this in production since some 4 years ago, meanwhile `closeio` has been running in production without this for as long. I guess now is a good time to merge and make `closeio` use this. I'll also update the `dialer` to use `master` once this is merged in, because `master` contains the redis-py 3 changes.

@thomasst you can't approve your own PR. Can I go ahead with this? I'll run a quick canary test against `closeio` tomorrow and merge if all goes well.

File: redis_hashring/__init__.py [hunk 1]
```diff
@@ -315,13 +324,11 @@ def poll(self):
 
         try:
             while True:
-                # Since Redis' listen method blocks, we use select to inspect the
-                # underlying socket to see if there is activity.
-                fileno = pubsub.connection._sock.fileno()
                 timeout = max(0, POLL_INTERVAL - (time.time() - last_heartbeat))
-                r, w, x = self._select([fileno], [], [], timeout)
-                if fileno in r:
-                    next(gen)
```
Comment (thomasst) [inline redis_hashring/__init__.py:324]: Above, is `gen = pubsub.listen()` still needed?

File: redis_hashring/__init__.py [hunk 2]
```diff
@@ -315,13 +324,11 @@ def poll(self):
...
+                message = pubsub.get_message(
```
Comment (thomasst) [inline redis_hashring/__init__.py:328]: Would it be an improvement if we turned this into a loop to avoid unnecessary updates?

```
has_messages = False
while pubsub.get_message(ignore_subscribe_messages=True, timeout=timeout):
    has_messages = True
if has_messages:
    self.update()
```

This can be implemented separately since I believe your change is equivalent, so we should merge this PR first before improving it.

**Comment by jkemp101** [comment]
I didn't remember we had this PR and was using it somewhere. I had to revert my attempt at a similar refactor https://github.com/closeio/redis-hashring/pull/6.

File: redis_hashring/__init__.py (hunk 1 above)
Comment (jpmelos) [reply redis_hashring/__init__.py:322]: Dropped it in https://github.com/closeio/redis-hashring/pull/3/commits/596f4aa1ef34e8f6b34b1624aec4f5b453771142.

File: redis_hashring/__init__.py (hunk 2 above)
Comment (jpmelos) [reply redis_hashring/__init__.py:320]: I think I like @jkemp101's implementation better, from https://github.com/closeio/redis-hashring/pull/5. I added it to this PR, please review to see if it accomplishes what you intended.

**Comment by jpmelos** [comment]
I ran tests for [`closeio`](https://app.circleci.com/pipelines/github/closeio/closeio/12565/workflows/cefef2dd-85bc-4a31-9299-9395214920df/jobs/220785) and [`dialer`](https://app.circleci.com/pipelines/github/closeio/dialer/284/workflows/b81159b3-c99a-4189-9c71-5ab334a49942/jobs/2811) with the [latest commit in this PR as of now](https://github.com/closeio/redis-hashring/pull/3/commits/506391fd333f97eb5d5ee8bb51864545febfc500) and the tests passed in both projects. So I think [whatever was causing the tests to block](https://github.com/closeio/redis-hashring/pull/6) when @jkemp101 tried to merge his changes is now solved. I think we can try to run this in `closeio` canary and see if we can put this in production there?

**Comment by jpmelos** [comment]
I verified that tests pass in some client projects, so I think whatever was the issue that made tests block when @jkemp101 tried to merge his changes is now solved.

File: redis_hashring/__init__.py [hunk 3]
```diff
@@ -5,7 +5,6 @@
 import random
 import operator
 import os
-import select
```
Comment (thomasst) [inline redis_hashring/__init__.py:8]: I think the original idea behind this is that you could use redis-hashring independently, either without gevent, or with non-monkey-patched gevent. Since this is an independent module, it's best not to make any assumptions on how the main application is set up.

By default it would use the system's select (which may or may not be gevent depending on whether you monkey-patched), but if you use `gevent_start()` it would always be gevent's `select` regardless of whether you monkey patched.

With this change you would break usage for users that prefer not to monkey-patch and use redis-hashring with gevent.

Comment (jpmelos) [reply redis_hashring/__init__.py:8]: Let me try to explain how I understand this to make sure I'm understanding it correctly.

**TL;DR:** The only change that belong in this PR is the `decode` change and the rest has to be reverted, right?

You're saying people are supposed to be allowed to use this project with `gevent` but without monkey-patching. This means if we stop using `listen` to use `get_message`, we will use the `sock.settimeout` under the hood and block the thread for that amount of time when waiting for the socket to have anything. `gevent` won't be aware of that blocking operation, since it didn't patch `socket`, and therefore the thread won't yield control.

So this means I need to revert any changes related to `get_message`, go back to using `listen` and `select.select`, and the only change that actually belongs in this PR is just the decode one. Right? This way, even if someone doesn't make `gevent` patch `select`, we will still be able to use `gevent.select` when polling our Redis pubsub subscription when running `poll` as a product of `gevent_start`.

Comment (jpmelos) [reply redis_hashring/__init__.py]: If that sounds right, then I undid the `get_message` change in https://github.com/closeio/redis-hashring/pull/3/commits/a0e7ab8a21a9b073fe42a82a309d5a87d742989a but still implemented draining the subscription channel before updating the node in https://github.com/closeio/redis-hashring/pull/3/commits/ccf380ca89e6984ada47b2e22eb8e0d9452b2c76. Let me know if that's not what you meant.

Comment (thomasst) [reply redis_hashring/__init__.py]: Hmm. Maybe for readability reasons it's fine as long as the README mentions that the functions need to be patched when using it in gevent mode _before_ importing redis-hashring. Sorry about the back and forth, after thinking about it one more time I don't think we need to support the niche case where you don't monkey patch, but we need to be aware of it and point it out. Also, in the raw select code the subscribe message would trigger an update, which is unnecessary.

It's not a strong opinion, if we leave it with `self._select` that's fine as well. In this case though I would add a comment saying why we're not using `get_message()` so that we don't get confused in the future.

Comment (jpmelos) [reply redis_hashring/__init__.py:8]: How does it look now?

File: README.rst [hunk 4]
```diff
@@ -171,7 +171,13 @@ gevent example
...
-replicas. An example app could look as follows:
+replicas.
+
+To use with `gevent`, make sure to have called `gevent.monkey.patch_all()`
+*before* importing *redis* and *redis-hashring* for the first time in the
+process.
+
+An example app could look as follows:
 
 .. code:: python
```
Comment (thomasst) [inline README.rst:182]: On top of the sample code:

```
import gevent.monkey

gevent.monkey.patch_all()
````

**Review by thomasst — COMMENTED** [review]
Looks good, you can merge.
